BrainPortal/lib/s3_connection.rb

```diff
@@ -19,52 +18,244 @@
+    else
+      return {}
+    end
+  end                                            
+      
+  # Copies an individual object from S3 back to a destination
+  def copy_object_from_bucket(srcObj,dest)
+    if object_exists?(srcObj)
+      resp = @resource.client.get_object(response_target: dest.to_s,
+                                        bucket: @bucket_name,
+                                        key: srcObj.to_s)
+    end
+  end
+  
+  
+  # Copies and sets up all of the directories to copy an entire path from S3 to dest
+  def copy_path_from_bucket(path, dest_head)
+    if object_exists?(path)
+      list_of_objects = list_objects_long(path)
+      list_of_objects.each do |x|
+        xStat = get_object_stats(x[:key])
+        next if xStat[:content_type] == 'application/x-directory'
+        trun_object_key = x[:key].dup.sub! "#{clean_starting_folder_path(File.dirname(path).to_s).to_s}/",''
+        fullDirName = File.join(dest_head,File.dirname(trun_object_key))
+        fullPathName = File.join(dest_head,trun_object_key)
+        FileUtils::mkdir_p fullDirName
+        copy_object_from_bucket(x[:key], fullPathName)
+      end
+    end
+  end
+  
+  # Copies a file to the bucket
+  def copy_file_to_bucket(srcFile, dest_head)
+    trun_object_key = File.basename(srcFile)
+    keyPath = File.join(dest_head,trun_object_key).to_s
+    resp = @resource.client.put_object(body: srcFile,
+                                       bucket: @bucket_name,
+                                       key: keyPath)
+  end
+  
+  # Copies a directory to the bucket, will recursively create directory in the path
+  def copy_directory_to_bucket(srcDir, dest_head)
+    if object_exists?(dest_head)
+      Dir.glob("#{srcDir}/**/*").each do |x|
+        next if File.directory?(x)
+        newDirName = "#{File.dirname(x).sub(File.dirname(srcDir),'')}"
+        new_dest_path = File.join(clean_starting_folder_path(dest_head).to_s,newDirName)
+        copy_file_to_bucket(x,new_dest_path)
+      end
+    end
   end
 
+  # Delete an object from the bucket
+  def delete_path_from_bucket(path)
+    path_clean = clean_starting_folder_path(path).to_s
+    if object_exists?(path_clean)
+      @resource.bucket(@bucket_name).objects({prefix: path_clean}).batch_delete!
+    end
+  end
+  
+  # Renames an objects path
+  def rename_object(srcPath,destPath)
+    src_path_clean = clean_starting_folder_path(srcPath).to_s
+    dest_path_clean = clean_starting_folder_path(destPath).to_s
+    if object_exists?(src_path_clean)
+      src_path_w_bucket = "/#{File.join(@bucket_name,src_path_clean)}"
+      resp = @resource.client.copy_object(bucket:@bucket_name,
+                                          copy_source: src_path_w_bucket,
+                                          key: dest_path_clean)
+      delete_path_from_bucket(src_path_clean)
+    end
+  end
+  
+  # Renames a Folder by renaming each object underneath
```

Comment:
There's no way to rename just a top folder?!?

Reply:
I am not sure what you mean, the bucket name should never be changed.  But you can change the name of any folder.